wallet/core/src/tx/generator/generator.rs

```diff
@@ -458,78 +512,129 @@ impl Generator {
             let utxo_entry_reference = if let Some(utxo_entry_reference) = self.get_utxo_entry(context, stage) {
                 utxo_entry_reference
             } else {
-                // UTXO sources are depleted, handle sweep processing
-                if self.inner.final_transaction_amount.is_none() {
-                    return self.finish_relay_stage_processing(context, stage, data);
+                // UTXO sources are depleted
+                if let Some(final_transaction) = &self.inner.final_transaction {
+                    // reject transaction
+                    return Err(Error::InsufficientFunds {
+                        additional_needed: final_transaction.value_with_priority_fee.saturating_sub(stage.aggregate_input_value),
+                        origin: "accumulator",
+                    });
                 } else {
-                    return Err(Error::InsufficientFunds);
+                    // finish sweep processing
+                    return self.finish_relay_stage_processing(context, stage, data);
                 }
             };
 
-            let UtxoEntryReference { utxo } = &utxo_entry_reference;
-
-            let input = TransactionInput::new(utxo.outpoint.clone().into(), vec![], input_sequence, self.inner.sig_op_count);
-            let input_amount = utxo.amount();
-            let input_mass = calc.calc_mass_for_input(&input) + self.inner.signature_mass_per_input;
-            input_sequence += 1;
-
-            // mass threshold reached, yield transaction
-            if data.aggregate_mass + input_mass + self.inner.standard_change_output_mass > MAXIMUM_STANDARD_TRANSACTION_MASS {
-                context.utxo_stash.push_back(utxo_entry_reference);
-                data.aggregate_mass += self.inner.standard_change_output_mass;
-                data.transaction_fees = self.calc_relay_transaction_relay_fees(&data);
-                stage.aggregate_fees += data.transaction_fees;
-                context.aggregate_fees += data.transaction_fees;
-                return Ok((DataKind::Node, data));
+            if let Some(node) = self.aggregate_utxo(context, calc, stage, &mut data, utxo_entry_reference) {
+                return Ok((node, data));
             }
 
+            if let Some(final_transaction) = &self.inner.final_transaction {
+                // try finish a stage or produce a final transaction with target value
+                // use basic condition checks to avoid unnecessary processing
+                if data.aggregate_mass > TRANSACTION_MASS_BOUNDARY_FOR_STAGE_INPUT_ACCUMULATION
+                    || (self.inner.final_transaction_priority_fee.sender_pays()
+                        && stage.aggregate_input_value >= final_transaction.value_with_priority_fee)
+                    || (self.inner.final_transaction_priority_fee.receiver_pays()
+                        && stage.aggregate_input_value >= final_transaction.value_no_fees - context.aggregate_fees)
```

Comment:
The original logic is correct. This is the case where we have inclusive `ReceiverPays` fees (which in overall also complicates the logic quite a bit TBH). When we operate in this mode, our boundary is the total amount as fees (for the entire context) become "inclusive".  Hence the boundary is `5 KAS - fees`. (and this is just an fn entrance boundary meant as a pre-condition to prevent extra computation).

Reply:
My suggestion performs the same check but w/o sub overflow possibility. I wasn't sure that we know that final_transaction.value_no_fees >= context.aggregate_fees in this context

Reply:
Thanks.  Didn't notice this. Fixing.